File: src/utils/problems.js
// This file will hold all the functions necessary to create random math problems for kids to solve
// This will produce random values for the problems
function digitHelper(digits) {
  const digitStringArray = ['1']
  let i = 1;
  while (i <= digits) {
    digitStringArray.push('0');
    i++
  }
  const num = Number(digitStringArray.join(''))
  const random = Math.floor(Math.random() * num)

  const x = String(random).length < digitStringArray.length - 1 ? num - random : random

  return x;
}

// Max digits: 6, Differing digits possible, Decimals TBD
function generateAdditionProblem(digits, secondDigits = null, decimalPlaces = null) {
  const x = digitHelper(digits);
  const y = secondDigits === null ? digitHelper(digits) : digitHelper(secondDigits);

  return {
    x,
    y,
    solution: x + y,
  }
}

// Max digits: 6, Differing digits possible, Decimals TBD
function generateSubtractionProblem(digits, secondDigits = null, decimalPlaces = null) {
  const x = digitHelper(digits);
  const y = secondDigits === null ? digitHelper(digits) : digitHelper(secondDigits);

  return {
    x,
    y,
    solution: x - y,
  }
}

// Max digits: 6, Differing digits possible, Decimals TBD
function generateMultiplicationProblem(digits, secondDigits = null, decimalPlaces = null) {
  const x = digitHelper(digits);
  const y = secondDigits === null ? digitHelper(digits) : digitHelper(secondDigits);

  return {
    x,
    y,
    solution: x * y,
  }
}

// Max digits: 6, Differing digits possible, Decimals TBD
function generateDivisionProblem(digits, secondDigits = null, decimalPlaces = null) {
  const x = digitHelper(digits);
  const y = secondDigits === null ? digitHelper(digits) : digitHelper(secondDigits);

  return {
    x,
    y,
    solution: x / y,
  }
}

// Max digits: 2, Max power: 10
function generateExponentProblem(digits, power) {
  let range;
  switch (digits) {
    case 1:
      range = 10;
      break;
    case 2:
      range = 100;
      break;
  }

  const randomX = Math.floor(Math.random() * range)

  return {
    x: randomX,
    power: power,
    solution: randomX ** power,
  }
}

console.log(generateAdditionProblem(2));
console.log(generateAdditionProblem(4, 3));
console.log(generateSubtractionProblem(3));
console.log(generateSubtractionProblem(3));
console.log(generateMultiplicationProblem(2));
console.log(generateMultiplicationProblem(2));
console.log(generateDivisionProblem(4));
console.log(generateDivisionProblem(4, 3));
console.log(generateExponentProblem(2, 10));

// Returns a range between the lowest possible number with the amount of digits specified and the largest possible number with the amount of digits specified.
// Example: digitHelper(3) will return a random value between 100 and 999, inclusively. 

File: src/utils/problems.py
from random import randrange

def getRandomIntInRange(start: int, stop: int) -> int:
  x = randrange(start, stop+1)
  return x

def getRangeEndpoints(digits) -> dict:
  if digits > 5: return "This function only allows up to 5 digits."
  match digits:
    case 1:
      digitRange = { "start": 1, "stop": 9 }
    case 2:
      digitRange = { "start": 10, "stop": 99 }
    case 3:
      digitRange = { "start": 100, "stop": 999 }
    case 4:
      digitRange = { "start": 1000, "stop": 9999 }
    case 5:
      digitRange = { "start": 10000, "stop": 99999 }        
  return digitRange


def generateAdditionProblem(digits, secondDigits=None) -> dict:
  xRange = getRangeEndpoints(digits)
  yRange = getRangeEndpoints(digits) if secondDigits == None else getRangeEndpoints(secondDigits)
  
  x = getRandomIntInRange(xRange["start"], xRange["stop"])
  y = getRandomIntInRange(yRange["start"], yRange["stop"])
  
  return { "x": x, "y": y, "solution": x + y }

def generateSubtractionProblem(digits, secondDigits=None, negativeAnswers=True) -> dict:
  if secondDigits > digits and negativeAnswers == False: return "This will only produce answers with negative values. Please change the digit amount of the second value or set negativeAnswers to true"
  xRange = getRangeEndpoints(digits)
  yRange = getRangeEndpoints(digits) if secondDigits == None else getRangeEndpoints(secondDigits)
  
  x = getRandomIntInRange(xRange["start"], xRange["stop"])
  y = getRandomIntInRange(yRange["start"], yRange["stop"])
    
  if negativeAnswers == False:
    while x - y < 0:
      x = getRandomIntInRange(xRange["start"], xRange["stop"])
      y = getRandomIntInRange(yRange["start"], yRange["stop"])
  return { "x": x, "y": y, "solution": x - y }

def generateMultiplicationProblem(digits, secondDigits=None, negativeAnswers=True) -> dict:
  xRange = getRangeEndpoints(digits)
  yRange = getRangeEndpoints(digits) if secondDigits == None else getRangeEndpoints(secondDigits)
  
  x = getRandomIntInRange(xRange["start"], xRange["stop"])
  y = getRandomIntInRange(yRange["start"], yRange["stop"])
  
  if negativeAnswers == False:
    while x - y < 0:
      x = getRandomIntInRange(xRange["start"], xRange["stop"])
      y = getRandomIntInRange(yRange["start"], yRange["stop"])
  
  return { "x": x, "y": y, "solution": x * y }

def generateDivisionProblem(digits, secondDigits=None, remainders=True) -> dict:
  if secondDigits > digits and remainders == False: return "This will only produce answers with remainders. Please change the digit amount of the second value or set remainders to true"
  xRange = getRangeEndpoints(digits)
  yRange = getRangeEndpoints(digits) if secondDigits == None else getRangeEndpoints(secondDigits)
  
  x = getRandomIntInRange(xRange["start"], xRange["stop"])
  y = getRandomIntInRange(yRange["start"], yRange["stop"])
  
  if remainders == False:
    while x % y != 0:
      x = getRandomIntInRange(xRange["start"], xRange["stop"])
      y = getRandomIntInRange(yRange["start"], yRange["stop"])
      
  return { "x": x, "y": y, "solution": x / y }

def run_x_times(x: int, func, *func_args) -> list:
  result = []
  i = 1
  while i <= x:
    val = func(*func_args)
    result.append(val)
    i += 1
  return result
